user/storage.md

```diff
@@ -0,0 +1,267 @@
+# Files and Data in the Cloud
+
+This page describes how files and data storage are handled in 2i2c Hubs.
+The high-level summary of recommendations is:
+- Use your home directory to store code, notebooks, and small data files (<1 GB)
+  for personal use
+- Use cloud object storage to store larger datasets and to share data across your team
+- Consider whether your project would benefit from other cloud-native data storage
+  solutions such as a database, data warehouse, or data lake
+
+:::{attribution}
+The following material was adapted from the
+[Pangeo Cloud User Guide](https://pangeo.io/cloud.html)
+:::
+
+## Your Home Directory
+
+Your notebook server is a linux "virtual machine" with its own filesystem.
+Your are not on a shared server; you are on your own private server.
+Your username is ``jovyan``, and your home directory is ``/home/jovyan``.
+This is the same for all users.
+
+Your home directory is intended only for notebooks, analysis scripts, and small datasets (< 1 GB).
+It is not an appropriate place to store large datasets.
+No one else can see or access the files your home directory.
+
+The easiest way to move files in and out of your home directory is via the JupyterLab web interface.
+Drag a file into the file browser to upload, and right-click to download back out.
+You can also open a terminal via the JupyterLab launcher and use this to ssh / scp / ftp to remote systems.
+However, you can’t ssh in!
+
+## The `shared` Directory
+
+All users have a directory called `shared` in their home directory.
+This is a *readonly* directory - anybody on the hub can *access* and *read from* the `shared` directory.
+The hub administrator may choose to distribute shared materials via this directory.
+The `shared` directory is not intended as a way for hub users to share data with each other.
+
+## Using Git / GitHub
```

Comment:
I know that @yuvipanda will not approve of the security practices recommended in this section. I am happy to replace this with something else that is more secure. We just need to give a clear answer to the question _how should I push to a github remote from 2i2c hubs_?

Reply:
@fperez has been dogfooding https://github.com/yuvipanda/github-app-user-auth for a few months now and is pretty happy with it. https://github.com/yuvipanda/github-app-user-auth/pull/8 also makes it work inline from within JupyterLab. 

I agree we should provide a clear answer to 'how do I push to github from a hub'. I can replace this section with something using https://github.com/yuvipanda/github-app-user-auth.

Reply:
> I can replace this section with something using https://github.com/yuvipanda/github-app-user-auth.

@yuvipanda, do you want to do that as part of this PR or maybe as a follow-up as @choldgraf suggested below?

Reply:
I've spent a bunch of time upgrading https://github.com/yuvipanda/gh-scoped-creds over the last couple days, I'll make that a PR to this PR.

Reply:
yeah, it'll be one per hub. So user workflow would be:

1. run 'gh-scoped-creds' (or `%ghscopedcreds`) from within the notebook
2. Grant this particular JupyterHub access
3. Manage which repos this particular repo has push access to.

The initial setup for hub admins isn't too big either, so I think this just becomes a 'feature' you can ask for when we make a hub.

Reply:
https://user-images.githubusercontent.com/30430/164171109-22a18151-a031-46b5-8e6f-7d0bc28cddc6.mp4

I made a quick demo of how this will look like. We can improve this over time of course. It lets users provide per-hub, per-repo permissions in a time-scoped manner. The second part where you're choosing the repo need not be done every 8h.

Reply:
I turned this into an announcement blog post https://blog.jupyter.org/securely-pushing-to-github-from-a-jupyterhub-3ee42dfdc54f that is a draft in the Jupyter blog. There's yet another big supply chain attack going on right now based on stolen GitHub credentials, so good timing for this :D

Reply:
I'm working on setting this up on the pangeo GCP hub. https://github.com/pangeo-data/pangeo-docker-images/pull/305 is currently waiting for review.

Reply:
I've test-deployed this onto https://staging.us-central1-b.gcp.pangeo.io - you can test it out by opening a notebook there, then:

```
%pip install gh-scoped-creds
import gh_scoped_creds
%ghscopedcreds
```

And following the prompts there. The `pip install` will go away once https://github.com/pangeo-data/pangeo-docker-images/pull/305 is merged.

https://github.com/2i2c-org/infrastructure/pull/1219 sets this config up and also adds docs on how to do it, so we can turn this on as a 'feature' for any new hub.

Reply:
And here's my PR to this PR replacing the ssh-key section! https://github.com/rabernat/2i2c-docs/pull/1

Reply:
This is an amazing solution.